common/enclave/cc_data.cpp

```diff
@@ -0,0 +1,225 @@
+
+    CATCH(b, host_parameters_.assign(host_parameters, host_parameters + hp_size));
+    COND2ERR(!b);
+
+    // build attested data
+    b = build_attested_data(attested_data);
+    COND2ERR(!b);
+
+    {
+        // get attestation parameters in proto
+        pb_istream_t is, sis;
+        pb_wire_type_t wt;
+        uint32_t tag;
+        bool eof;
+
+        is = pb_istream_from_buffer(attestation_parameters, ap_size);
+        COND2LOGERR(!pb_decode_tag(&is, &wt, &tag, &eof), "cannot decode tag");
+        COND2LOGERR(wt != PB_WT_STRING, "unexpected type");
+        COND2LOGERR(tag != attestation_AttestationParameters_parameters_tag, "unexpected tag");
+
+        COND2LOGERR(!pb_make_string_substream(&is, &sis), "substream error");
+
+        {
+            // retrieve parameters in substream
+            std::string b64_ap_s;
+            CATCH(b, b64_ap_s.assign(sis.bytes_left, '0'));
+            COND2ERR(!b);
+
+            b = pb_read(&sis, (pb_byte_t*)b64_ap_s.c_str(), b64_ap_s.length());
+            COND2LOGERR(!b, "cannot read field");
+
+            attestation_parameters_s = base64_decode(b64_ap_s);
+        }
+
+        pb_close_string_substream(&is, &sis);
+    }
+
+    // init attestation
+    b = init_attestation(
+        (uint8_t*)attestation_parameters_s.c_str(), attestation_parameters_s.length());
+    COND2LOGERR(!b, "cannot init attestation");
+
+    // get attestation
+    CATCH(b, attestation.resize(attestation_max_length));
+    COND2ERR(!b);
+
+    b = get_attestation(attested_data.data(), attested_data.size(), attestation.data(),
+        attestation_max_length, &attestation_length);
+    COND2LOGERR(!b, "cannot get attestation");
+
+    {
+        // build credentials (Attested_Data || Attestation)
+        pb_ostream_t ostream;
+        ostream = pb_ostream_from_buffer(credentials, credentials_max_size);
+        {
+            pb_ostream_t ostream_any;
+            ByteArray buffer;
+            // NOTE: buffer size should be adapted
+            CATCH(b, buffer.resize(attested_data.size() + 1024));
+            COND2ERR(!b);
+
+            {
+                // serialize the Any type
+                ostream_any = pb_ostream_from_buffer(buffer.data(), buffer.size());
+
+                COND2ERR(
+                    !pb_encode_tag(&ostream_any, PB_WT_STRING, google_protobuf_Any_type_url_tag));
+                // NOTE: the url type string is necessary,
+                //       and the type after last '/' must match the serialized message type
+                std::string s("github.com/fpc/fpc.Attested_Data");
+                COND2ERR(
+                    !pb_encode_string(&ostream_any, (const unsigned char*)s.c_str(), s.length()));
+
+                COND2ERR(!pb_encode_tag(&ostream_any, PB_WT_STRING, google_protobuf_Any_value_tag));
+                COND2ERR(!pb_encode_string(&ostream_any, (const unsigned char*)attested_data.data(),
+                    attested_data.size()));
+            }
```

Comment:
Did you double check that the credential and attestedData protos can also deserialized outside the enclave? I didn't see this in a test, maybe I missed it. This will happen at ercc and during ecc validation.

Maybe you can use this piece of code for a quick check in `enclave_stub.go`

```go
if err := proto.Unmarshal(credentialBytes, &credentials); err != nil {
	return errors.Wrap(err, "invalid credential bytes")
}

var attestedData protos.AttestedData
if err := ptypes.UnmarshalAny(credentials.SerializedAttestedData, &attestedData); err != nil {
	return errors.Wrap(err, "invalid attested data message")
}

// check some of the values
```

Reply:
I had removed the related debug code during cleanup.
Anyway, added again.